src/implementations/astroport/pool.rs

```diff
@@ -0,0 +1,422 @@
+use astroport_core::asset::PairInfo;
+use astroport_core::factory::PairType;
+use astroport_core::querier::{query_supply, query_token_precision};
+use astroport_core::U256;
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{
+    to_binary, Addr, CosmosMsg, Decimal, Env, QuerierWrapper, QueryRequest, Response, StdError,
+    StdResult, WasmMsg, WasmQuery,
+};
+use cosmwasm_std::{Deps, Event, Uint128};
+use cw20::Cw20ExecuteMsg;
+use cw_asset::{Asset, AssetInfo, AssetInfoBase, AssetList};
+
+use astroport_core::asset::Asset as AstroAsset;
+use astroport_core::asset::AssetInfo as AstroAssetInfo;
+use astroport_core::pair::{
+    Cw20HookMsg, ExecuteMsg as PairExecMsg, PoolResponse, QueryMsg, QueryMsg as PairQueryMsg,
+    SimulationResponse,
+};
+
+use crate::traits::Pool;
+use crate::CwDexError;
+
+use super::helpers::{
+    adjust_precision, astro_asset_info_to_cw_asset_info, compute_current_amp, compute_d,
+    cw_asset_to_astro_asset, query_pair_config, AstroAssetList, N_COINS,
+};
+
+#[cw_serde]
+pub struct AstroportPool {
+    pub pair_addr: Addr,
+    pub lp_token_addr: Addr,
+    pub pair_type: PairType,
+}
+
+impl AstroportPool {
+    pub fn new(deps: Deps, pair_addr: Addr) -> StdResult<Self> {
+        let pair_info =
+            deps.querier.query_wasm_smart::<PairInfo>(pair_addr.clone(), &PairQueryMsg::Pair {})?;
+
+        // Validate pair type. We only support XYK and stable swap pools
+        match pair_info.pair_type {
+            PairType::Custom(_) => Err(StdError::generic_err("Custom pair type is not supported")),
+            _ => Ok(()),
+        }?;
+
+        Ok(Self {
+            pair_addr,
+            lp_token_addr: pair_info.liquidity_token,
+            pair_type: pair_info.pair_type,
+        })
+    }
+
+    pub fn query_lp_token_supply(&self, querier: &QuerierWrapper) -> StdResult<Uint128> {
+        query_supply(querier, self.lp_token_addr.to_owned())
+    }
+
+    pub fn get_pool_liquidity_impl(&self, querier: &QuerierWrapper) -> StdResult<PoolResponse> {
+        querier.query::<PoolResponse>(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: self.pair_addr.to_string(),
+            msg: to_binary(&QueryMsg::Pool {})?,
+        }))
+    }
+
+    /// Math for LP shares calculation when providing liquidity to an Astroport constant product pool.
+    ///
+    /// Copied from the astroport XYK pool implementation here:
+    /// https://github.com/astroport-fi/astroport-core/blob/7bedc6f27e59ef8b921a0980be9bc30c4aab7459/contracts/pair/src/contract.rs#L297-L434
+    pub fn xyk_simulate_provide_liquidity(
```

Comment:
Why this pub but not stableswap equivalent?

Reply:
```suggestion
    fn xyk_simulate_provide_liquidity(
```